.get panics with Output<string> but works with string

Was able to repro this issue with the following code on the version the reporter was on

```typescript
import * as pulumi from "@pulumi/pulumi";
import * as azure from "@pulumi/azure-native";

const resourceGroup = new azure.resources.ResourceGroup("resourceGroup");
const lbId = pulumi.secret(resourceGroup.id)

export const works = lbId.apply((id) => azure.resources.ResourceGroup.get("thing", id, undefined)) // works
export const notWorks = azure.resources.ResourceGroup.get("thingtoo", lbId, undefined) // doesn't work
```

```
Do you want to perform this update? yes
Updating (dev)

View Live: https://app.pulumi.com/kyle-pulumi-corp/azure-lb/dev/updates/27

     Type                                     Name           Status      Info
     pulumi:pulumi:Stack                      azure-lb-dev   running..   read azure-native:resources:ResourceGroup thing
 +   └─ azure-native:resources:ResourceGroup  resourceGroup  created     
panic: fatal: An assertion has failed

goroutine 295 [running]:
github.com/pulumi/pulumi/sdk/v3/go/common/util/contract.failfast(...)
	/home/kdixler/go/src/github.com/pulumi/pulumi/sdk/go/common/util/contract/failfast.go:23
github.com/pulumi/pulumi/sdk/v3/go/common/util/contract.Assert(...)
	/home/kdixler/go/src/github.com/pulumi/pulumi/sdk/go/common/util/contract/assert.go:26
github.com/pulumi/pulumi/sdk/v3/go/common/resource/plugin.(*provider).Read(0x1f0a228?, {0xc00032c0f0?, 0x10?}, {0x0?, 0x0?}, 0x7f9562d15cc8?, 0x7f958ca7ea68?)
	/home/kdixler/go/src/github.com/pulumi/pulumi/sdk/go/common/resource/plugin/provider_plugin.go:809 +0xb54
github.com/pulumi/pulumi/pkg/v3/resource/deploy.(*ReadStep).Apply(0xc0016d9560, 0x48?)
	/home/kdixler/go/src/github.com/pulumi/pulumi/pkg/resource/deploy/step.go:636 +0xcc
github.com/pulumi/pulumi/pkg/v3/resource/deploy.(*stepExecutor).executeStep(0xc001566300, 0x0?, {0x1f0a228, 0xc0016d9560})
	/home/kdixler/go/src/github.com/pulumi/pulumi/pkg/resource/deploy/step_executor.go:302 +0x1e2
github.com/pulumi/pulumi/pkg/v3/resource/deploy.(*stepExecutor).executeChain(0xc001566300, 0xc000513728?, {0xc0017febf0?, 0x1, 0x0?})
	/home/kdixler/go/src/github.com/pulumi/pulumi/pkg/resource/deploy/step_executor.go:252 +0xe5
github.com/pulumi/pulumi/pkg/v3/resource/deploy.(*stepExecutor).worker.func1()
	/home/kdixler/go/src/github.com/pulumi/pulumi/pkg/resource/deploy/step_executor.go:414 +0xd9
created by github.com/pulumi/pulumi/pkg/v3/resource/deploy.(*stepExecutor).worker
	/home/kdixler/go/src/github.com/pulumi/pulumi/pkg/resource/deploy/step_executor.go:411 +0x2b1
```

on SDK v3.86.0 it with master complains with

```
% pulumi up
Previewing update (dev)

View in Browser (Ctrl+O): https://app.pulumi.com/kyle-pulumi-corp/azure-lb/dev/previews/a2455bd4-0e9e-46e4-8a8d-def6fb70b7f0

     Type                                     Name          Plan     Info
     pulumi:pulumi:Stack                      azure-lb-dev           
     └─ azure-native:resources:ResourceGroup  thingtoo               1 error


Diagnostics:
  azure-native:resources:ResourceGroup (thingtoo):
    error: Expected an ID for urn:pulumi:dev::azure-lb::azure-native:resources:ResourceGroup::thingtoo

```

--- a/changelog/pending/20231019--sdk-nodejs--nodejs-now-supports-unknown-resource-ids.yaml
+++ b/changelog/pending/20231019--sdk-nodejs--nodejs-now-supports-unknown-resource-ids.yaml
@@ -1,0 +1,4 @@
+changes:
+- type: fix
+  scope: sdk/nodejs
+  description: Nodejs now supports unknown resource IDs.

--- a/sdk/nodejs/runtime/resource.ts
+++ b/sdk/nodejs/runtime/resource.ts
@@ -244,13 +244,13 @@ export function readResource(
     opts: ResourceOptions,
     sourcePosition?: SourcePosition,
 ): void {
-    const id: Input<ID> | undefined = opts.id;
-    if (!id) {
+    if (!opts.id) {
         throw new Error("Cannot read resource whose options are lacking an ID value");
     }
+    const id: Promise<Input<ID>> = output(opts.id).promise(true);
 
     const label = `resource:${name}[${t}]#...`;
-    log.debug(`Reading resource: id=${Output.isInstance(id) ? "Output<T>" : id}, t=${t}, name=${name}`);
+    log.debug(`Reading resource: t=${t}, name=${name}`);
 
     const monitor = getMonitor();
     const resopAsync = prepareResource(label, res, parent, true, false, props, opts);
@@ -258,7 +258,7 @@ export function readResource(
     const preallocError = new Error();
     debuggablePromise(
         resopAsync.then(async (resop) => {
-            const resolvedID = await serializeProperty(label, id, new Set(), { keepOutputValues: false });
+            const resolvedID = await serializeProperty(label, await id, new Set(), { keepOutputValues: false });
             log.debug(
                 `ReadResource RPC prepared: id=${resolvedID}, t=${t}, name=${name}` +
                     (excessiveDebugOutput ? `, obj=${JSON.stringify(resop.serializedProps)}` : ``),

--- a/sdk/nodejs/tests/runtime/langhost/cases/014.read_resource/index.js
+++ b/sdk/nodejs/tests/runtime/langhost/cases/014.read_resource/index.js
@@ -10,7 +10,7 @@ let inputs = {
     d: undefined,
 };
 
-let res = new pulumi.CustomResource("test:read:resource", "foo", inputs, { id: "abc123" });
+let res = new pulumi.CustomResource("test:read:resource", "foo", inputs, { id: pulumi.secret("abc123") });
 res.id.apply((id) => assert.strictEqual(id, "abc123"));
 res.urn.apply((urn) => assert.strictEqual(urn, "test:read:resource::foo"));
 res.a.apply((a) => assert.strictEqual(a, inputs.a)); // same as input

